x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java

```diff
@@ -221,20 +224,40 @@ public void getRoles(User user, Authentication authentication, ActionListener<Ro
 
         final Authentication.AuthenticationType authType = authentication.getAuthenticationType();
         if (authType == Authentication.AuthenticationType.API_KEY) {
-            apiKeyService.getRoleForApiKey(authentication, ActionListener.wrap(apiKeyRoleDescriptors -> {
-                final List<RoleDescriptor> descriptors = apiKeyRoleDescriptors.getRoleDescriptors();
-                if (descriptors == null) {
-                    roleActionListener.onFailure(new IllegalStateException("missing role descriptors"));
-                } else if (apiKeyRoleDescriptors.getLimitedByRoleDescriptors() == null) {
-                    buildAndCacheRoleFromDescriptors(descriptors, apiKeyRoleDescriptors.getApiKeyId() + "_role_desc", roleActionListener);
-                } else {
-                    buildAndCacheRoleFromDescriptors(descriptors, apiKeyRoleDescriptors.getApiKeyId() + "_role_desc",
-                        ActionListener.wrap(role -> buildAndCacheRoleFromDescriptors(apiKeyRoleDescriptors.getLimitedByRoleDescriptors(),
-                            apiKeyRoleDescriptors.getApiKeyId() + "_limited_role_desc", ActionListener.wrap(
-                                limitedBy -> roleActionListener.onResponse(LimitedRole.createLimitedRole(role, limitedBy)),
-                                roleActionListener::onFailure)), roleActionListener::onFailure));
-                }
-            }, roleActionListener::onFailure));
+            if (authentication.getVersion().onOrAfter(Version.V_7_9_0)) {
+                getOrBuildRoleForApiKey(authentication, false, ActionListener.wrap(
+                    role -> {
+                        if (role == Role.EMPTY) {
```

Comment:
When can this be true?

Reply:
When a list of role descriptor is empty, the role building method [buildRoleFromDescriptors](https://github.com/elastic/elasticsearch/blob/v7.8.0/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java#L360-L363) returns `Role.EMPTY`.

Also when an API key does not have its own role descriptors, its document will have `{}` for the `role_descriptors` field, which in turn will be converted to an empty list of `RoleDescriptor`.

Reply:
Got it! I have to check the code every single time to tell if the `role_descriptors` field or the `limited_by_role_descriptors` can be empty, the meanings are reversed in `ApiKeyService#getRoleForApiKey`.

Since `limitedByRole` can never be `null` I would be building that first, for a minor branching simplification:
```
         final Authentication.AuthenticationType authType = authentication.getAuthenticationType();
         if (authType == Authentication.AuthenticationType.API_KEY) {
             if (authentication.getVersion().onOrAfter(Version.V_7_9_0)) {
-                buildAndCacheRoleForApiKey(authentication, false, ActionListener.wrap(
-                    role -> {
-                        if (role == Role.EMPTY) {
-                            buildAndCacheRoleForApiKey(authentication, true, roleActionListener);
-                        } else {
-                            buildAndCacheRoleForApiKey(authentication, true, ActionListener.wrap(
-                                limitedByRole -> roleActionListener.onResponse(
-                                    limitedByRole == Role.EMPTY ? role : LimitedRole.createLimitedRole(role, limitedByRole)),
-                                roleActionListener::onFailure
+                buildAndCacheRoleForApiKey(authentication, true, ActionListener.wrap(
+                        limitedByRole -> {
+                            buildAndCacheRoleForApiKey(authentication, false, ActionListener.wrap(
+                                    role -> roleActionListener.onResponse(
+                                            role == Role.EMPTY ? limitedByRole : LimitedRole.createLimitedRole(role, limitedByRole)),
+                                    roleActionListener::onFailure
                             ));
-                        }
-                    },
-                    roleActionListener::onFailure
+                        },
+                        roleActionListener::onFailure
                 ));
             } else {
                 apiKeyService.getRoleForApiKey(authentication, ActionListener.wrap(apiKeyRoleDescriptors -> {
```

Feel free to follow this suggestion or not.

Reply:
You may be right. But I am not entirely sure when it comes to derived API keys. A derived keys will have limited by role descriptors from parent API key's roles. Since we now [zero out](https://github.com/elastic/elasticsearch/blob/7.8/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java#L514) API key's roles during authentication, we may end up having an empty limited by role descriptors as well. I have not tested, but I feel it is probably safer to keep the code as is for now.